src/main/java/org/csanchez/jenkins/plugins/kubernetes/SelfRegisteredSlaveRetentionStrategy.java

```diff
@@ -0,0 +1,111 @@
+package org.csanchez.jenkins.plugins.kubernetes;
+
+import com.google.common.annotations.VisibleForTesting;
+import hudson.model.Computer;
+import hudson.model.Node;
+import hudson.slaves.Cloud;
+import hudson.slaves.CloudSlaveRetentionStrategy;
+import hudson.slaves.OfflineCause;
+import hudson.util.TimeUnit2;
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import jenkins.model.Jenkins;
+import org.apache.commons.lang.StringUtils;
+import org.jvnet.localizer.Localizable;
+import org.jvnet.localizer.ResourceBundleHolder;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ */
+public class SelfRegisteredSlaveRetentionStrategy extends CloudSlaveRetentionStrategy {
+
+    private static final Logger LOGGER = Logger.getLogger(SelfRegisteredSlaveRetentionStrategy.class.getName());
+
+    /**
+     * The resource bundle reference
+     */
+    private final static ResourceBundleHolder HOLDER = ResourceBundleHolder.get(Messages.class);
+
+    private final String cloudName;
+    private final String namespace;
+    private final int maxMinutesIdle;
+
+    public SelfRegisteredSlaveRetentionStrategy(String cloudName, String namespace, int maxMinutesIdle) {
+        this.cloudName = cloudName;
+        this.namespace = namespace;
+        this.maxMinutesIdle = maxMinutesIdle;
+    }
+
+    @Override
+    protected long getIdleMaxTime() {
+        return TimeUnit2.MINUTES.toMillis(maxMinutesIdle);
+    }
+
+    @Override
+    protected long checkCycle() {
+        return 1;
+    }
+
+    @Override
+    protected void kill(Node n) throws IOException {
+        LOGGER.info("Node " + n.getNodeDescription() + " will be killed now");
+        terminate(n);
+    }
+
+    private void terminate(Node n) throws IOException {
```

Comment:
I did some refactoring already, as per Mihails's comments above, please see my next set of changes coming soon